### ISSUE TITLE
[Feature Request] API Access for User Account Token

### ISSUE THREAD

**Issue by Adron** [drone/drone#2129]
Currently we're using Drone.io to setup environments for use in building a wide array of applications using various stacks. The technology stacks we're using range from Go, Java, .NET Core, Python, JavaScript/Node.js and others. When we setup these environments, the entire ecosystem - from individual environments, user accounts, and related collateral - are spun up via an automated system created with Terraform, Bash, and some other tools. When these tools set everything up various pieces of state are passed from execution to execution to set all of these elements up. This automated part of the process has one huge feature we'd love to be able to have, to complete more automation around Drone.io. That feature is the ability at the API level to get a _user account token_ so that we can take additional steps via the API as that user the _user account token_ is assigned to.

**_NOTE:_** I did review [Github Issue 1812](https://github.com/drone/drone/issues/1812), which was a somewhat similar yet tangential feature request. But I wanted to be specific that I'm looking for this capability to use Drone.io in an internal environment. It may still be in a public cloud service like AWS or GCP, but it wouldn't be accessible to Github or some of the other services. So a robot user account as referred in [issue 1812](https://github.com/drone/drone/issues/1812) wouldn't really work for this, _but_ this feature could prospective lend itself to building a robot feature as specified in [issue 1812](https://github.com/drone/drone/issues/1812). As per the other comments too it looks like that request and this feature could be implemented in conjunction.

I've created this issue, not just to propose the feature but to see if it would be something that the Drone.io team would like to have added to the code base? If so, I'm happy to dive in and get code written to enable this and submit a PR, but wanted to check with the team to insure this is wanted or not.

Cheers,
@Adron

**Comment by bradrydzewski** [comment]
If you would like to automatically add users and generate keys for the purpose of automation, I would recommend interacting directly with the database to create the user.

This is the minimal statement required to insert a user

```
INSERT INTO users (
 user_login
,user_token
,user_secret
,user_expiry
,user_email 
,user_avatar
,user_active
,user_admin 
,user_hash 
) values (?, '', '', '', '', '', true, false, ?)
```

The only two required fields (identified by the ?) are the user login (github username) and a random hash that is 32 characters long. The code for generating the has is [here](https://github.com/drone/drone/blob/1c01cd36128b0685b27cab79dcba3fc404e238a0/server/users.go#L68-L70).

The random hash is used to sign the JWT token with the HS256 signing algorithm. Note that we use JWT because it is a standard with implementations in pretty much any language, which makes it very easy to generate drone tokens as long as you know the hash.

Note that the JWT token expects two claims:

```
{
  "text": "octocat",
  "type": "user"
}
```

This is an example token (invalid signature) that you can paste into [jwt.io](https://jwt.io/) to get a feel for the token structure. You can use this website to create a valid drone token, provided you paste the user hash in the secret textbox.

`
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ0eXBlIjoidXNlciIsInRleHQiOiJvY3RvY2F0In0.1PthkhwQquQqoyneGX72VtQ2SyIoZTdbnRqcBNxAtIE
`

So in summary it should be quite easy to automate the creation of user accounts and generate tokens during your setup process, but this will require a small amount of custom code. The general position I take on this sort of stuff is that we should be building external tools vs building capabilities into drone directly.

To that end perhaps this could just be automated with a script that is included in your automation script? Or perhaps this merits an official library or toolkit that can be used with various automation engines.

This is a sample Go program that should get you started

```Go
package main

import (
	"database/sql"
	"encoding/base32"
	"flag"
	"log"

	"github.com/dgrijalva/jwt-go"
	"github.com/gorilla/securecookie"
	_ "github.com/mattn/go-sqlite3"
)

func main() {
	var conn string
	var user string

	flag.StringVar(&conn, "conn", "", "database connection string")
	flag.StringVar(&user, "user", "", "username (i.e. octocat)")
	flag.Parse()

	hash := base32.StdEncoding.EncodeToString(
		securecookie.GenerateRandomKey(32),
	)

	db, err := sql.Open("sqlite3", conn)
	if err != nil {
		log.Fatal(err)
	}

	_, err = db.Exec(stmt, user, hash)
	if err != nil {
		log.Fatal(err)
	} else {
		log.Printf("user account created [%s]", user)
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims{
		StandardClaims: jwt.StandardClaims{},
		Type:           "user",
		Text:           user,
	})
	secret, err := token.SignedString([]byte(hash))
	if err != nil {
		log.Fatal(err)
	}

	log.Printf("user token generated [%s]", secret)
}

type claims struct {
	jwt.StandardClaims
	Type string `json:"type"`
	Text string `json:"text"`
}

const stmt = `
INSERT INTO users (
 user_login
,user_token
,user_secret
,user_expiry
,user_email
,user_avatar
,user_active
,user_admin
,user_hash
) values (?,'','','','','',true,false,?)
`

```

**Comment by tallpauley** [comment]
@bradrydzewski Is there any "guarantee" of stability for the user db table? I would love to write a Vault secret backend plug-in for drone service accounts using this method, but I don't want to maintain this if the table schema is likely to change often.

**Comment by bradrydzewski** [comment]
@tallpauley I do not expect we will ever off a guarantee of database stability since that is considered an internal implementation detail. We will instead strive for a guarantee of stability with the external facing integration points, including the yaml and REST API.

In terms of vault integration, you may be interested in our vault backend http://docs.drone.io/vault-secrets/

**Comment by tallpauley** [comment]
@bradrydzewski makes sense, I just thought I'd ask.

Yeah, not interested in Drone fetching vault secrets. Interested in Vault provisioning drone service accounts, but I think I'll drop this if there will be no stable way to interact with Drone directly to create service accounts. Sounds like the only supported solution for provisioning service accounts is creating the service accounts in the source control system.

**Comment by bradrydzewski** [comment]
> Sounds like the only supported solution for provisioning service accounts is creating the service accounts in the source control system.

This is correct. For service accounts you would create a [machine user](https://developer.github.com/v3/guides/managing-deploy-keys/#machine-users) in your source code management system (e.g. GitHub) and then register the machine user as a drone administrator, assuming administrative access is required.

The primary reason is that Drone does not have its own user management; it is delegated to the source code management system. So having a user account that is not linked to GitHub, for example, would be a paradigm shift for Drone.

**Comment by bradrydzewski** [comment]
@StevenACoffman if I understand your use case correctly, this is already supported by using [secrets]( http://docs.drone.io/manage-secrets/). You would provide a secret to your pipeline, which you could then use in your regression tests. If you have any questions, feel free to break this out into a separate thread in our [support forum](discourse.drone.io)

**Comment by bradrydzewski** [comment]
We are going to be able to accommodate this use case upcoming release. You will be able to bootstrap the system with a user account (a real user, or a machine user) and you will be able to supply the user api token (if not, one is auto-generated). This token can then be used to make API calls.

```
DRONE_USER_CREATE=username:octocat,machine:false,admin:true,token:55f24eb3d61ef6ac5e83d550178638dc
```

**Comment by bradrydzewski** [comment]
implemented in 1.0.0-rc.1
